### ISSUE TITLE
[BUG]: whkd connection is broken while using komorebi 

### ISSUE THREAD

**Issue by Jaehaks** [LGUG2Z/whkd#49]
**Describe the bug**
whkd connection is broken after I use some command repeatively
Restarting komorebi solve this issue, but this problem invokes very often

**To Reproduce**
1. `komorebic stop --whkd`
2. do something, (go cross workspace, shift layouts) 
3. finally, whkd doesn't work

**Expected behavior**
whkd doesn't loose its control

**Screenshots and Videos**

https://github.com/user-attachments/assets/063ccfd3-b2c4-4708-90d2-eff0c37cd1f7

**Operating System**
OS Name:                   Microsoft Windows 10 Pro
OS Version:                10.0.19045 N/A Build 19045

**Additional context**

In particular, if you have any other AHK scripts or software running that handle any aspect of window management, manipulation or hotkey binding

**Comment by LGUG2Z** [comment]
This is fixed in the latest nightly build and will be part of v0.1.31 : https://github.com/LGUG2Z/komorebi/commit/7f0b54c35e932e5fc753d5e24738de7303763dcc

**Comment by Jaehaks** [comment]
it has same issue even though using nightly version. 
@LGUG2Z

**Comment by alex-ds13** [comment]
> it has same issue even though using nightly version. @LGUG2Z

Can you run the following commands and post the result, please?
```ps1
komorebic --version
```
```ps1
where.exe komorebic
```
```ps1
where.exe komorebi
```

**Comment by Jaehaks** [comment]
@alex-ds13 

Yeah, I erased komorebi from scop list.
there are no other execution file.

Now I tested In my home PC, It seems to work.
but In PC at work, whkd loose control the same with before even though the komorebi and whkd version are same with home PC. 

I tried 3 times and whkd loose control after swap window several times (9~10) in all case. 
I'll test again tomorrow

**Comment by Jaehaks** [comment]
@alex-ds13 

After test on work PC,  The whkd issue exists yet, I tested 5 times iteratively

version is correct. I use `komorebic nightly` and `whkd v0.2.2`
I repeated flipping horizontal layout using shortcut very quickly.  
sometimes connection of whkd is broken yet. 
After broken, `komorebic.exe` is remained in windows task manager. and `komorebi.exe` and two `whkd.exe` too.
`komorebi.exe` and `whkd.exe` is gone after I typed `komorebic stop --whkd` in cmd. but `komorebic.exe` is remained 

this is result of `komorebic log` at the time when whkd loose control

```powershell
2024-11-20T23:11:33.215429Z  INFO process_command{FlipLayout(Horizontal)}:flip_layout{layout_flip=Horizontal}:update_focused_workspace{follow_focus=false trigger_focus=false}: komorebi::window_manager: updating
2024-11-20T23:11:33.227549Z  INFO process_command{FlipLayout(Horizontal)}: komorebi::process_command: processed
2024-11-20T23:11:33.230645Z  INFO process_event{event="TitleUpdate" winevent="ObjectNameChange" hwnd=593124}: komorebi::process_event: processed: (hwnd: 593124, title: C:\Users\USER\scoop\shims\whkd.exe, exe: whkd.exe, class: ConsoleWindowClass)
2024-11-20T23:11:33.351862Z  INFO process_event{event="TitleUpdate" winevent="ObjectNameChange" hwnd=593124}: komorebi::process_event: processed: (hwnd: 593124, title: C:\Users\USER\scoop\shims\whkd.exe - komorebic  flip-layout horizontal, exe: whkd.exe, class: ConsoleWindowClass)
2024-11-20T23:11:33.369155Z  INFO process_command{FlipLayout(Horizontal)}:flip_layout{layout_flip=Horizontal}:komorebi::window_manager: flipping layout
2024-11-20T23:11:33.369562Z  INFO process_command{FlipLayout(Horizontal)}:flip_layout{layout_flip=Horizontal}:update_focused_workspace{follow_focus=false trigger_focus=false}: komorebi::window_manager: updating
2024-11-20T23:11:33.382088Z  INFO process_command{FlipLayout(Horizontal)}: komorebi::process_command: processed
2024-11-20T23:11:33.385310Z  INFO process_event{event="TitleUpdate" winevent="ObjectNameChange" hwnd=593124}: komorebi::process_event: processed: (hwnd: 593124, title: C:\Users\USER\scoop\shims\whkd.exe, exe: whkd.exe, class: ConsoleWindowClass)
2024-11-20T23:11:33.486997Z  INFO process_event{event="TitleUpdate" winevent="ObjectNameChange" hwnd=593124}: komorebi::process_event: processed: (hwnd: 593124, title: C:\Users\USER\scoop\shims\whkd.exe - komorebic  flip-layout horizontal, exe: whkd.exe, class: ConsoleWindowClass)
2024-11-20T23:11:33.508675Z  INFO process_command{FlipLayout(Horizontal)}:flip_layout{layout_flip=Horizontal}: komorebi::window_manager: flipping layout
# connection is broken
```

**Comment by Jaehaks** [comment]
@alex-ds13 

environment is same with home PC 

```powershell
2024/11/21 22:34:32 C:\Users\USER
@ >> komorebic --version
komorebic 0.1.31
branch:master
commit_hash:818ec1c6
build_time:2024-11-17 00:54:18 +00:00
build_env:rustc 1.82.0 (f6e511eec 2024-10-15),stable-x86_64-pc-windows-msvc

2024/11/21 22:34:36 C:\Users\USER
@ >> where komorebic
C:\Users\USER\.config\installed\komorebi-nightly-x86_64-pc-windows-msvc\komorebic.exe

2024/11/21 22:34:48 C:\Users\USER
@ >> where  komorebi
C:\Users\USER\.config\installed\komorebi-nightly-x86_64-pc-windows-msvc\komorebi.exe
```

**Comment by alex-ds13** [comment]
> @alex-ds13
> 
> environment is same with home PC
> 
> ```powershell
> 2024/11/21 22:34:32 C:\Users\USER
> @ >> komorebic --version
> komorebic 0.1.31
> branch:master
> commit_hash:818ec1c6
> build_time:2024-11-17 00:54:18 +00:00
> build_env:rustc 1.82.0 (f6e511eec 2024-10-15),stable-x86_64-pc-windows-msvc
> 
> 2024/11/21 22:34:36 C:\Users\USER
> @ >> where komorebic
> C:\Users\USER\.config\installed\komorebi-nightly-x86_64-pc-windows-msvc\komorebic.exe
> 
> 2024/11/21 22:34:48 C:\Users\USER
> @ >> where  komorebi
> C:\Users\USER\.config\installed\komorebi-nightly-x86_64-pc-windows-msvc\komorebi.exe
> ```

Try uninstalling all komorebi versions you have on your system completely and then after there is no komorebi or komorebic, install the nightly version again!
I've just see a user have some weird issues (with a newer version having the behavior of an older version...) when going between versions after `winget upgrade` and completely uninstalling it and then install again solved it

**Comment by Jaehaks** [comment]
@alex-ds13 

It is weird.... I reinstalled 3 times but the results are same. 
I'll test after official `komorebi v0.1.31` is released.

When komorebi starts, is this normal response? 
 
```powershell
2024/11/22 09:29:31 C:\Users\USER\Desktop
@ >> komorebic start --whkd
Start-Process 'komorebi.exe' -WindowStyle hidden
Waiting for komorebi.exe to start...Started!

if (!(Get-Process whkd -ErrorAction SilentlyContinue))
{
  Start-Process whkd -WindowStyle hidden
}


Thank you for using komorebi!

# Sponsorship
* Become a sponsor https://github.com/sponsors/LGUG2Z - $5/month makes a big difference
* Leave a tip https://ko-fi.com/lgug2z - An alternative to GitHub Sponsors
* Subscribe to https://youtube.com/@LGUG2Z - Development videos, feature previews and release overviews

# Community
* Join the Discord https://discord.gg/mGkn66PHkx - Chat, ask questions, share your desktops
* Explore the Awesome Komorebi list https://github.com/LGUG2Z/awesome-komorebi - Projects in the komorebi ecosystem

# Documentation
* Read the docs https://lgug2z.github.io/komorebi - Quickly search through all komorebic commands
```

**Comment by alex-ds13** [comment]
> After broken, komorebic.exe is remained in windows task manager. and komorebi.exe and two whkd.exe too.

There shouldn't be 2 `whkd.exe` instances running. If you start komorebi with `komorebic --whkd` and then stop it without the `--whkd` flag then the `whkd` will stay open and after that you start komorebi again with the `--whkd` flag there will be two instances of `whkd` this might be the cause of the behavior. Try to force stop one of the instances on TaskManager and see if it still happens.

> this is result of `komorebic log` at the time when whkd loose control
> 
> ```powershell
> 2024-11-20T23:11:33.215429Z  INFO process_command{FlipLayout(Horizontal)}:flip_layout{layout_flip=Horizontal}:update_focused_workspace{follow_focus=false trigger_focus=false}: komorebi::window_manager: updating
> 2024-11-20T23:11:33.227549Z  INFO process_command{FlipLayout(Horizontal)}: komorebi::process_command: processed
> 2024-11-20T23:11:33.230645Z  INFO process_event{event="TitleUpdate" winevent="ObjectNameChange" hwnd=593124}: komorebi::process_event: processed: (hwnd: 593124, title: C:\Users\USER\scoop\shims\whkd.exe, exe: whkd.exe, class: ConsoleWindowClass)
> 2024-11-20T23:11:33.351862Z  INFO process_event{event="TitleUpdate" winevent="ObjectNameChange" hwnd=593124}: komorebi::process_event: processed: (hwnd: 593124, title: C:\Users\USER\scoop\shims\whkd.exe - komorebic  flip-layout horizontal, exe: whkd.exe, class: ConsoleWindowClass)
> 2024-11-20T23:11:33.369155Z  INFO process_command{FlipLayout(Horizontal)}:flip_layout{layout_flip=Horizontal}:komorebi::window_manager: flipping layout
> 2024-11-20T23:11:33.369562Z  INFO process_command{FlipLayout(Horizontal)}:flip_layout{layout_flip=Horizontal}:update_focused_workspace{follow_focus=false trigger_focus=false}: komorebi::window_manager: updating
> 2024-11-20T23:11:33.382088Z  INFO process_command{FlipLayout(Horizontal)}: komorebi::process_command: processed
> 2024-11-20T23:11:33.385310Z  INFO process_event{event="TitleUpdate" winevent="ObjectNameChange" hwnd=593124}: komorebi::process_event: processed: (hwnd: 593124, title: C:\Users\USER\scoop\shims\whkd.exe, exe: whkd.exe, class: ConsoleWindowClass)
> 2024-11-20T23:11:33.486997Z  INFO process_event{event="TitleUpdate" winevent="ObjectNameChange" hwnd=593124}: komorebi::process_event: processed: (hwnd: 593124, title: C:\Users\USER\scoop\shims\whkd.exe - komorebic  flip-layout horizontal, exe: whkd.exe, class: ConsoleWindowClass)
> 2024-11-20T23:11:33.508675Z  INFO process_command{FlipLayout(Horizontal)}:flip_layout{layout_flip=Horizontal}: komorebi::window_manager: flipping layout
> # connection is broken
> ```

How exactly did you know that the connection was broken at that exact time?

> After I quit the komorebic.exe, whkd shortcut works well, but this situation will happens again soon. (about 3~4 attempts)

Does this mean that when the issue happens, if you force stop the `komorebic` on TaskManager, then komorebi starts working normally again?


### Further questions:
Is your home PC also running Windows 10?

Are you able to build komorebi from source so you can add some debug lines where I tell you to, to see if we can pin point exactly what is going on here? This doesn't seem to be the old issue where the komorebi socket would get stuck and commands wouldn't work anymore, that issue used to happen to me and since the fix it never happened again. Recently I moved to an older version without the fix and within a couple of hours the issue happened again. So the fix seems to work. Also with that issue, even if you forced stopped `komorebic` that wouldn't fix it, since the problem was on `komorebi`. What you are facing might be something else...

**Comment by Jaehaks** [comment]
> There shouldn't be 2 `whkd.exe` instances running. If you start komorebi with `komorebic --whkd` and then stop it without the `--whkd` flag then the `whkd` will stay open and after that you start komorebi again with the `--whkd` flag there will be two instances of `whkd` this might be the cause of the behavior. Try to force stop one of the instances on TaskManager and see if it still happens.

It is the first time I have noticed that whkd has only one instance in Windows Task Manager. 
When I execute komorebi using komorebi start --whkd, there is one instance of komorebic and two instances of whkd always.
This result doesn't come from not terminating komorebi without the --whkd flag.
The both `whkd` always are terminated together when I quit one of them from windows manager.
Can you see that `whkd` has only one process in task manager in you PC?

When I stop komorebi, this result is followed
```powershell
2024/11/24 09:16:24 C:\Users\USER\.config\nvim  main
@ >> komorebic stop --whkd

Stop-Process -Name:whkd -ErrorAction SilentlyContinue
```

> How exactly did you know that the connection was broken at that exact time?

I use `komorebic log` after staring komorebi. 
When the problem occurs, It doesn't show any next log.

> Does this mean that when the issue happens, if you force stop the komorebic on TaskManager, then komorebi starts working normally again?

You are right. Although the problem does occur again, it still works fairly well.

> Is your home PC also running Windows 10?

Yeah, PC in home and work are windows 10 

> Are you able to build komorebi from source so you can add some debug lines where I tell you to, to see if we can pin point exactly what is going on here? This doesn't seem to be the old issue where the komorebi socket would get stuck and commands wouldn't work anymore, that issue used to happen to me and since the fix it never happened again. Recently I moved to an older version without the fix and within a couple of hours the issue happened again. So the fix seems to work. Also with that issue, even if you forced stopped komorebic that wouldn't fix it, since the problem was on komorebi. What you are facing might be something else...

I am not an expert in Rust, but I'll try my best as soon as possible.

**Comment by alex-ds13** [comment]
> Can you see that whkd has only one process in task manager in you PC?

Yes. To me there is only one instance of `whkd` as it should.

That `komorebic` process that you always see, is because of you running `komorebic log` that creates an instance of `komorebic` that is always running until you stop it. Can you try using komorebi normally without running `komorebic log` and make sure there is only on `whkd` instance. Don't worry about the logs you can access them later any way since they are stored on a file, usually on the `%LOCALAPPDATA%\TEMP\' folder...

See if running komorebi without that `komorebic log` running still shows that behavior. Also if this happens again try running the command `komorebic state` to see if it prints anything.

**Comment by Jaehaks** [comment]
> That komorebic process that you always see, is because of you running komorebic log that creates an instance of komorebic that is always running until you stop it

i am sorry to wrong notice... there are one instance of "komorebi" and two instances of whkd always. 

komorebic process is shown while only i send tiling command. like swap, log, etc... 

but it is also weird. i thought two whkd are always normal. because it happened about 5 month ago. what i knew the komorebi

i will check the `komorebic state`

**Comment by Jaehaks** [comment]
@alex-ds13 

The result of `komorebic state` is 
```json
{
  "monitors": {
    "elements": [
      {
        "id": 65539,
        "name": "DISPLAY2",
        "device": "LGD057B",
        "device_id": "LGD057B-4&340123b&0&UID265988",
        "size": {
          "left": -1920,
          "top": 360,
          "right": 1920,
          "bottom": 1080
        },
        "work_area_size": {
          "left": -1920,
          "top": 360,
          "right": 1920,
          "bottom": 1040
        },
        "work_area_offset": null,
        "window_based_work_area_offset": null,
        "window_based_work_area_offset_limit": 1,
        "workspaces": {
          "elements": [
            {
              "name": "I",
              "containers": {
                "elements": [],
                "focused": 0
              },
              "monocle_container": null,
              "maximized_window": null,
              "floating_windows": [],
              "layout": {
                "Default": "VerticalStack"
              },
              "layout_rules": [],
              "layout_flip": null,
              "workspace_padding": 0,
              "container_padding": 0,
              "latest_layout": [],
              "resize_dimensions": [],
              "tile": true,
              "apply_window_based_work_area_offset": true,
              "window_container_behaviour": null,
              "float_override": null
            }
          ],
          "focused": 0
        },
        "workspace_names": {}
      },
      {
        "id": 65537,
        "name": "DISPLAY3",
        "device": "PHLC170",
        "device_id": "PHLC170-4&340123b&0&UID198147",
        "size": {
          "left": 0,
          "top": 0,
          "right": 2560,
          "bottom": 1440
        },
        "work_area_size": {
          "left": 0,
          "top": 0,
          "right": 2560,
          "bottom": 1400
        },
        "work_area_offset": null,
        "window_based_work_area_offset": null,
        "window_based_work_area_offset_limit": 1,
        "workspaces": {
          "elements": [
            {
              "name": null,
              "containers": {
                "elements": [
                  {
                    "id": "Z1Z0VwTlBgmgzVs2hIPpn",
                    "windows": {
                      "elements": [
                        {
                          "hwnd": 133320,
                          "title": "Command Prompt - komorebic  state",
                          "exe": "WindowsTerminal.exe",
                          "class": "CASCADIA_HOSTING_WINDOW_CLASS",
                          "rect": {
                            "left": 3,
                            "top": 3,
                            "right": 1274,
                            "bottom": 1394
                          }
                        }
                      ],
                      "focused": 0
                    }
                  },
                  {
                    "id": "w98S5YoD4d5RjH2QGwdnR",
                    "windows": {
                      "elements": [
                        {
                          "hwnd": 133012,
                          "title": "add non-integer initial trigger by Jaehaks · Pull Request #14 · dedukun/relative-motions.yazi - Brave",
                          "exe": "brave.exe",
                          "class": "Chrome_WidgetWin_1",
                          "rect": {
                            "left": 1283,
                            "top": 3,
                            "right": 1274,
                            "bottom": 1394
                          }
                        }
                      ],
                      "focused": 0
                    }
                  }
                ],
                "focused": 0
              },
              "monocle_container": null,
              "maximized_window": null,
              "floating_windows": [],
              "layout": {
                "Default": "BSP"
              },
              "layout_rules": [],
              "layout_flip": null,
              "workspace_padding": 0,
              "container_padding": 0,
              "latest_layout": [
                {
                  "left": 3,
                  "top": 3,
                  "right": 1274,
                  "bottom": 1394
                },
                {
                  "left": 1283,
                  "top": 3,
                  "right": 1274,
                  "bottom": 1394
                }
              ],
              "resize_dimensions": [
                null,
                null
              ],
              "tile": true,
              "apply_window_based_work_area_offset": true,
              "window_container_behaviour": null,
              "float_override": null
            }
          ],
          "focused": 0
        },
        "workspace_names": {}
      }
    ],
    "focused": 1
  },
  "is_paused": false,
  "resize_delta": 50,
  "new_window_behaviour": "Append",
  "float_override": false,
  "cross_monitor_move_behaviour": "Insert",
  "unmanaged_window_operation_behaviour": "Op",
  "work_area_offset": null,
  "focus_follows_mouse": null,
  "mouse_follows_focus": false,
  "has_pending_raise_op": false
}
```

**Comment by alex-ds13** [comment]
But did you get that `State` while the issue was happening?

**Comment by Jaehaks** [comment]
> But did you get that `State` while the issue was happening?

I checked the result of `komorebic state` about 3 stage.
1) right after starting komorebi
2) after happening issue (komorebi is locked)
3) after restoring normal state (rmove `komorebic.exe` process) 

Fortunately. komorebi locking happends in the same window state with the one right after starting komorebi. so It can be easy to compare it.
The reults of them are same except time when I send command "komorebic state" 

When I send some komorebic command using shortcut after komorebi is locked,  this command doesn't work in that situation. But After removing the `komorebic.exe` process which is locking, the rest of commands are executed until the list of the pending commands is empty

**Comment by alex-ds13** [comment]
I still didn't understand if when is locked and you can't use keybinds, are you able to run a command on the terminal, like `komorebic state` and that works?! Or does it not work until you remove `komorebic.exe` process?

**Comment by Jaehaks** [comment]
@alex-ds13 

Input `komorebic` command manually works fine in terminal although keybinding is broken. That is why I thought it seems like whkd issue. I terminate the komorebi process with `komorebic stop --whkd` to quit both `komorebi` and `whkd` process together when The locking issue occurs.

I found something, If I install `whkd` using `scoop`, windows manager has 2 `whkd` process when I start it. Both `komorebic start --whkd` and `whkd` only. 
If I download `whkd` from github release, it works with only one `whkd` process although they are the same version `v0.2.2`

The `komorebic.exe` process locking issue is remained still although I use both `komorebi` and `whkd` recent release version. (both have one process each)

**Comment by alex-ds13** [comment]
When the keybinds are locked can you also run the command on terminal: `komorebic flip-layout vertical` (or `horizontal` depending on your layout...)?

Also can you share your `whkdrc` file?

**Comment by Jaehaks** [comment]
@alex-ds13 

Yeah , I mean all komorebic commands work well besides `start` and `stop`.

`whkdrc` is 
```
.shell cmd

# Reload whkd configuration
alt + o                 : taskkill /f /im whkd.exe && start /b whkd # if shell is cmd
# alt + o                 : taskkill /f /im whkd.exe; Start-Process whkd -WindowStyle hidden # if shell is pwsh / powershell
alt + shift + o         : komorebic reload-configuration

# App shortcuts - these require shell to be pwsh / powershell
# The apps will be focused if open, or launched if not open
# alt + f                 : if ($wshell.AppActivate('Firefox') -eq $False) { start firefox }
# alt + b                 : if ($wshell.AppActivate('Chrome') -eq $False) { start chrome }

alt + q                 : komorebic close
alt + shift + q         : komorebic stop --whkd

# Focus windows
alt + left               : komorebic focus left
alt + down               : komorebic focus down
alt + up                 : komorebic focus up
alt + right              : komorebic focus right
# alt + shift + oem_4     : komorebic cycle-focus previous # oem_4 is [
# alt + shift + oem_6     : komorebic cycle-focus next # oem_6 is ]

# Move windows
alt + ctrl + left		: komorebic move left
alt + ctrl + down		: komorebic move down
alt + ctrl + up		    : komorebic move up
alt + ctrl + right		: komorebic move right
# alt + shift + return    : komorebic promote

# Stack windows
alt + shift + left      : komorebic stack left
alt + shift + down  	: komorebic stack down
alt + shift + up    	: komorebic stack up
alt + shift + right 	: komorebic stack right
# alt + shift + oem_1 	: komorebic unstack # oem_1 is ;
alt + ctrl + f 		: komorebic unstack # oem_1 is ;
alt + shift + oem_1 	: komorebic unstack-all # oem_1 is ;
alt + shift + oem_4 	: komorebic cycle-stack previous # oem_4 is [
alt + shift + oem_6 	: komorebic cycle-stack next # oem_6 is ]
# move other monitor
alt + ctrl + r         : komorebic cycle-move-to-monitor next

# Resize
alt + shift + oem_plus  : komorebic resize-axis horizontal increase
alt + shift + oem_minus : komorebic resize-axis horizontal decrease
# alt + shift + oem_plus  : komorebic resize-axis vertical increase
# alt + shift + oem_minus : komorebic resize-axis vertical decrease

# Manipulate windows
	# monocle : use full screen but not maximize. it shows border
	# maximize : use full screen such as maximize button. it doesn't show border
	# float : turn off managing of specific windows
	# pause : turn off managing of all windows
alt + w                 : komorebic toggle-float
alt + ctrl + m         : komorebic toggle-monocle
# alt + ctrl + m 		: komorebic toggle-maximize
alt + ctrl + n         : komorebic minimize
alt + p                 : komorebic toggle-pause

# Window manager options
# alt + shift + r         : komorebic retile

# Layouts
alt + ctrl + h          : komorebic flip-layout horizontal
alt + ctrl + k          : komorebic flip-layout vertical

# Workspaces
ctrl + 1                 : komorebic focus-workspace 0
ctrl + 2                 : komorebic focus-workspace 1
ctrl + 3                 : komorebic focus-workspace 2
ctrl + 4                 : komorebic focus-workspace 3
ctrl + 5                 : komorebic focus-workspace 4
ctrl + 6                 : komorebic focus-workspace 5
ctrl + 7                 : komorebic focus-workspace 6
ctrl + 8                 : komorebic focus-workspace 7

# Move windows across workspaces
alt + ctrl + 1         : komorebic move-to-workspace 0
alt + ctrl + 2         : komorebic move-to-workspace 1
alt + ctrl + 3         : komorebic move-to-workspace 2
alt + ctrl + 4         : komorebic move-to-workspace 3
alt + ctrl + 5         : komorebic move-to-workspace 4
alt + ctrl + 6         : komorebic move-to-workspace 5
alt + ctrl + 7         : komorebic move-to-workspace 6
alt + ctrl + 8         : komorebic move-to-workspace 7


```

**Comment by alex-ds13** [comment]
Oh I didn't realize when you said you used `cmd`, that you also used it on `whkd`. Ok two things to try:
1- First stop all `komorebi` and `whkd` instances
2- Start `komorebi` with command `komorebic start` (don't use the `whkd` flag!)
3- Then start `whkd` by simply running `whkd.exe` on your terminal (this will make it so you might see some errors that might e happening on `whkd`

After doing this and checking if `whkd` prints any errors when this happens, I would then try to change your `.shell` from `cmd` to `pwsh` (if you have it) or at least `powershell` and see if this issue still happens when using one of those.

**Comment by Jaehaks** [comment]
@alex-ds13 

ok i'll do 
if i set the `.shell` to `pwsh`, do i have to send command `komorebic start` in pwsh shell? not in command prompt?

**Comment by Jaehaks** [comment]
@alex-ds13 

> 1- First stop all komorebi and whkd instances
2- Start komorebi with command komorebic start (don't use the --whkd flag!)
3- Then start whkd by simply running whkd.exe on your terminal (this will make it so you might see some errors that might be happening on whkd

I try it with setting `.shell` to `pwsh`. but the same locking issue occurs.

The phenomenon is like this :
1) After locking, `komorebic.exe` process is remained in windows task manager.
2) `whkd` process in terminal shows what the command is asked in every input although komorebi doesn't work because of locking problem. the whkd command log is updated even though it doesn't work
```
komorebic flip-layout vertical
komorebic flip-layout vertical
komorebic flip-layout vertical
komorebic flip-layout horizontal
komorebic flip-layout horizontal
komorebic flip-layout horizontal
```
3) I terminated the `whkd` process, and then the `komorebi.exe` and `komorebic.exe` are remained. 
After I restart `whkd` in cmd prompt, komorebi does work. I think komorebi create another `komorebic.exe` process to execute komorebic command, separate from `komorebic.exe` which was locking in before. 
4) return to 1) step, After another locking issue occurs, there are one `komorebi.exe` and two `komorebic.exe` process in windows task manager.

**Comment by alex-ds13** [comment]
@LGUG2Z I'm out of ideas here! This is a different issue than the one we fixed.

**TLDR:**
 - Sometimes the keybinds stop working
 - Some `komorebic.exe` is lingering on TaskManager when this happens
   - If the user then tries to run `komorebic` commands from terminal those still work *(This is one of the main differences, with the bug we fixed it would be `komorebi` that would lock the `process_commands` thread but here it seems to be working still)*
   - If the user force stops `whkd` and starts `whkd` again, then it starts working again, although the previous `komorebic.exe` process still shows on task manager
   - If instead of stopping `whkd` the user kills that `komorebic.exe` process, again everything starts working normally...

I'm getting out of ideas of how to debug this more...

**My suspicions:**
1) It might be some bug on `whkd` or `komorebic`
2) It might be some bug on `komorebi-client` on `send_message`, maybe the `write_all` might be getting stuck like it happened on the read part and we might be able to fix it by adding a `write_timeout` to the stream...
3) It might be the cause of the read block we were noticing on `komorebi` that was fixed with the `read_timeout`. So `komorebi` might be fixed and continues to work, but this instance of `komorebic` that maybe was the culprit of that entire bug in the first place continues to be stuck.

However I find it hard that it is option 3, since I've never experienced any issue anymore and I used to face it quite often, so if it was that I should also notice this happening to me. This probably applies to option 2 as well.

Maybe you have some ideas?!

**Comment by alex-ds13** [comment]
@Jaehaks Another thing you could try is use AHK instead of WHKD and see if the issue still happens...

**Comment by Jaehaks** [comment]
@alex-ds13 

I try to use ahk, but it has the same issue.
And also I tried with your download version. is has the same issue.  It seems that it isn't `whkd` issue...

I found something.
When I send `komorebic flip-layout horizontal` using shortcut repeately. this issue invokes. and there is one `komorebic.exe` process is remained in windows task manager.
After then, When I send `komorebic flip-layout vertical` using shorcut, it works. 
It means that the only command about `flip-layout horizontal` is locked. 
Keep going with `komorebic flip-layout vertical` using shortcut, It stucks too. and there are two `komorebic.exe` process are remained in winodws task manager.

In this situation., `komorebic flip-layout horizontal` or `komorebic flip-layout vertical` using command works but only shorcut is locked.

**Comment by alex-ds13** [comment]
> And also I tried with your download version.

Just to be sure can you do a `komorebic --version` on a terminal? Just so we are certain you were really using the new version...

**Comment by alex-ds13** [comment]
> When I send `komorebic flip-layout horizontal` using shortcut repeately. this issue invokes. and there is one `komorebic.exe` process is remained in windows task manager. After then, When I send `komorebic flip-layout vertical` using shorcut, it works. It means that the only command about `flip-layout horizontal` is locked. Keep going with `komorebic flip-layout vertical` using shortcut, It stucks too. and there are two `komorebic.exe` process are remained in winodws task manager.
> 
> In this situation., `komorebic flip-layout horizontal` or `komorebic flip-layout vertical` using command works but only shorcut is locked.

@LGUG2Z This almost feels as if that specific keybind got unregistered or that it's callback is locked...

**Comment by Jaehaks** [comment]
@alex-ds13 

```powershell
@ >> komorebic --version
komorebic 0.1.31
branch:
commit_hash:847a28a5
build_time:2024-11-27 22:39:36 +00:00
build_env:rustc 1.82.0 (f6e511eec 2024-10-15),stable-x86_64-pc-windows-msvc
```

**Comment by alex-ds13** [comment]
> It seems different with [ce258ca](https://github.com/LGUG2Z/komorebi/pull/1151/commits) what you tagged right?

Now thats fine! The github actions attributes it a different commit, which is the one you have. So that means that the option 2 I mentioned above is not the cause of this issue.
